solr/test-framework/src/java/org/apache/solr/common/cloud/ClusterStateUtil.java

```diff
@@ -164,86 +72,83 @@ public static boolean waitForAllReplicasNotLive(ZkStateReader zkStateReader, int
-    Slice[] slices;
-    slices = zkStateReader.getClusterState().getCollection(collection).getActiveSlicesArr();
-    int liveAndActive = 0;
-    for (Slice slice : slices) {
-      for (Replica replica : slice.getReplicas()) {
-        boolean live = zkStateReader.getClusterState().liveNodesContain(replica.getNodeName());
-        boolean active = replica.getState() == Replica.State.ACTIVE;
-        if (live && active) {
-          liveAndActive++;
-        }
-      }
-    }
-    return liveAndActive;
+    ClusterState clusterState = zkStateReader.getClusterState();
+    var liveNodes = clusterState.getLiveNodes();
+    var state = clusterState.getCollection(collection);
+    return (int)
+        replicasOfActiveSlicesStream(state)
+            .filter(replica -> liveAndActivePredicate(replica, liveNodes))
+            .count();
+  }
+
+  public static Stream<Replica> replicasOfActiveSlicesStream(DocCollection collectionState) {
+    return collectionState.getActiveSlices().stream()
+        .map(Slice::getReplicas)
+        .flatMap(Collection::stream);
   }
 
   public static boolean waitForLiveAndActiveReplicaCount(
       ZkStateReader zkStateReader, String collection, int replicaCount, int timeoutInMs) {
-    long timeout =
-        System.nanoTime() + TimeUnit.NANOSECONDS.convert(timeoutInMs, TimeUnit.MILLISECONDS);
-    boolean success = false;
-    while (!success && System.nanoTime() < timeout) {
-      success = getLiveAndActiveReplicaCount(zkStateReader, collection) == replicaCount;
+    return waitFor(
+        zkStateReader,
+        timeoutInMs,
+        collection,
+        (liveNodes, state) ->
+            replicasOfActiveSlicesStream(state)
+                    .filter(replica -> liveAndActivePredicate(replica, liveNodes))
+                    .count()
+                == replicaCount);
+  }
+
+  public static boolean waitFor(
+      ZkStateReader zkStateReader,
+      long timeoutMs,
+      String collection,
+      CollectionStatePredicate predicate) {
+    if (collection != null) {
+      try {
+        zkStateReader.waitForState(collection, timeoutMs, TimeUnit.MILLISECONDS, predicate);
+        return true;
+      } catch (InterruptedException e) {
+        Thread.currentThread().interrupt();
+        throw new SolrException(ErrorCode.SERVER_ERROR, "Interrupted");
+      } catch (TimeoutException e) {
+        return false;
+      }
+    }
+    // TODO first timeout on clusterState existing (add method to ZkStateReader?) then
 
-      if (!success) {
-        try {
-          Thread.sleep(TIMEOUT_POLL_MS);
-        } catch (InterruptedException e) {
-          Thread.currentThread().interrupt();
-          throw new SolrException(ErrorCode.SERVER_ERROR, "Interrupted");
+    final long timeoutAtNs =
+        System.nanoTime() + TimeUnit.NANOSECONDS.convert(timeoutMs, TimeUnit.MILLISECONDS);
+    while (System.nanoTime() < timeoutAtNs) {
```

Comment:
Maybe change into a `do...while` so that:
- It runs at least once even if `timeoutMs` is 0 (or non 0 + an improbable GC at the wrong moment)
- It doesn't `sleep` then fails directly without checking again if `predicate` matches when the timeout is exceeded